core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala

```diff
@@ -735,7 +749,12 @@ object SparkSubmit extends CommandLineUtils {
     }
 
     try {
-      mainMethod.invoke(null, childArgs.toArray)
+      if (isSparkApp) {
+        val envvars = Map[String, String]() ++ sys.env
+        mainMethod.invoke(null, childArgs.toArray, childSparkConf, envvars.toMap)
```

Comment:
In that case it might be worth it to add a check in `SparkLauncher` to throw an exception in case env variables are set, and the app is started in a thread.